### PR TITLE
Refactor out fresh name prefix extraction logic

### PR DESCRIPTION
**Pull request by densh**
As discussed in https://github.com/densh/scala/commit/0bcb907a25fd0688653c497e2d83a7c2c05e8f8e#diff-9b7190b44c5da63343ed67c2bfa6d043R194
1. refactor out FreshNameExtractor out of Quasiquotes cake into SymbolTable (can’t put it outside due to the fact that names are path-dependent)
2. add optional parameter to the fresh name creator to cover additional qq$ prefix needed for quasiquotes
3. add unit tests

review by @retronym

### CODE REVIEW

**Comment by JamesIry** [comment]
I think @gkossakowski has also spent some time de-caking things so mentioning him in case he wants to take a look at this.

**Comment by densh** [comment]
ping @retronym

File: src/reflect/scala/reflect/internal/FreshNames.scala [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * Copyright 2005-2013 LAMP/EPFL
+ */
+
+package scala
+package reflect
+package internal
+
+import scala.reflect.internal.util.FreshNameCreator
+
+trait FreshNames { self: SymbolTable =>
```
Comment (gkossakowski) [inline src/reflect/scala/reflect/internal/FreshNames.scala]: How about minimizing the dependencies? I believe you need to depend only on `StdNames` and `Names`, right?

Comment (densh) [inline src/reflect/scala/reflect/internal/FreshNames.scala]: True, just `Names` should be enough.

https://github.com/densh/scala/commit/d6ef83a2d75292e44e165a438594c035ed1b1330

**Comment by retronym** [comment]
LGTM. Thanks for the improved unit tests and docs.

**Pull request merged**